client/blocks/stripe-utils/normalize.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * @typedef {import('./type-defs').StripePaymentItem} StripePaymentItem
+ * @typedef {import('./type-defs').StripeShippingOption} StripeShippingOption
+ * @typedef {import('./type-defs').StripeShippingAddress} StripeShippingAddress
+ * @typedef {import('./type-defs').StripePaymentResponse} StripePaymentResponse
+ * @typedef {import('@woocommerce/type-defs/registered-payment-method-props').PreparedCartTotalItem} CartTotalItem
+ * @typedef {import('@woocommerce/type-defs/cart').CartShippingOption} CartShippingOption
+ * @typedef {import('@woocommerce/type-defs/shipping').ShippingAddress} CartShippingAddress
+ * @typedef {import('@woocommerce/type-defs/billing').BillingData} CartBillingAddress
+ */
+
+/**
+ * Normalizes incoming cart total items for use as a displayItems with the
+ * Stripe api.
+ *
+ * @param {CartTotalItem[]} cartTotalItems CartTotalItems to normalize
+ * @param {boolean}         pending        Whether to mark items as pending or
+ *                                         not
+ *
+ * @return {StripePaymentItem[]} An array of PaymentItems
+ */
+const normalizeLineItems = ( cartTotalItems, pending = false ) => {
+	return cartTotalItems
+		.map( ( cartTotalItem ) => {
+			return cartTotalItem.value
+				? {
+						amount: cartTotalItem.value,
+						label: cartTotalItem.label,
+						pending,
+				  }
+				: false;
+		} )
+		.filter( Boolean );
+};
+
+/**
+ * Normalizes incoming cart shipping option items for use as shipping options
+ * with the Stripe api.
+ *
+ * @param {CartShippingOption[]}  shippingOptions An array of CartShippingOption items.
+ *
+ * @return {StripeShippingOption[]}  An array of Stripe shipping option items.
+ */
+const normalizeShippingOptions = ( shippingOptions ) => {
+	const rates = shippingOptions[ 0 ].shipping_rates;
+	return rates.map( ( rate ) => {
+		return {
+			id: rate.rate_id,
+			label: rate.name,
+			detail: rate.description,
+			amount: parseInt( rate.price, 10 ),
+		};
+	} );
+};
+
+/**
+ * Normalize shipping address information from stripe's address object to
+ * the cart shipping address object shape.
+ *
+ * @param {StripeShippingAddress} shippingAddress Stripe's shipping address item
+ *
+ * @return {CartShippingAddress} The shipping address in the shape expected by
+ * the cart.
+ */
+const normalizeShippingAddressForCheckout = ( shippingAddress ) => {
+	const address = {
+		first_name: shippingAddress.recipient
+			.split( ' ' )
+			.slice( 0, 1 )
+			.join( ' ' ),
+		last_name: shippingAddress.recipient
+			.split( ' ' )
+			.slice( 1 )
+			.join( ' ' ),
+		company: '',
+		address_1:
+			typeof shippingAddress.addressLine[ 0 ] === 'undefined'
```

Comment:
We could also rewrite this to be `shippingAddress.addressLine[ 0 ] === undefined`, which is a little bit shorter and (IMO) easier to read.

That said, @nerrad do we have any guarantees that `addressLine` actually exists on the `shippingAddress` object? If there is no `addressLine` property, this will throw an error that we don't catch AFAICT.

Reply:
`shippingAddress` is coming from Stripe events. I assumed we had guarantees for the shape here when coming from Stripe. If in your experience, you've found that not to be the case, then ya additional type guards would be good here.

Reply:
>  I assumed we had guarantees for the shape here when coming from Stripe

There may very well be. I'll consult the documentation.

That said, extra guards probably won't hurt, but I think we can trust whatever the Stripe docs tell us 🙂 . I was worried this was something we created ourselves, and whether we made sure this wasn't undefined.

Reply:
Based on [Stripe's own TS definitions](https://github.com/stripe/stripe-js/blob/db51d9576a66ff0de4326aeac19dc73934935bc8/types/stripe-js/payment-request.d.ts#L355-L408) and [their documentation](https://stripe.com/docs/js/payment_request/events/on_shipping_address_change#payment_request_on_shipping_address_change-handler-shippingAddress) I do think we need more checks for the normalization functions.

I'm going to dig into the specifics a bit, and go through this in a new review. I'll mark this conversation as resolved for now, because the discussion on each case will be moved to its own thread in the new review.